cli/file_watcher.rs

```diff
@@ -28,51 +90,55 @@ async fn error_handler(watch_future: WatchFuture) {
 }
 
 pub async fn watch_func<F>(
-  watch_paths: &[PathBuf],
+  paths: &[PathBuf],
   closure: F,
 ) -> Result<(), AnyError>
 where
   F: Fn() -> WatchFuture,
 {
+  let (_watcher, receiver) = new_watcher(paths)?;
+  let mut debounce = Debounce::new(receiver, DEBOUNCE_TIMEOUT_MS);
   loop {
     let func = error_handler(closure());
     let mut is_file_changed = false;
     select! {
-      _ = file_watcher(watch_paths) => {
-          is_file_changed = true;
-          info!(
-            "{} File change detected! Restarting!",
-            colors::intense_blue("Watcher")
-          );
-        },
-      _ = func => { },
-    };
+      _ = debounce.next() => {
+        is_file_changed = true;
+        info!(
+          "{} File change detected! Restarting!",
+          colors::intense_blue("Watcher"),
+        );
+      },
+      _ = func => {},
+    }
```

Comment:
An else `sleep_for(DEBOUNCE_TIMEOUT_MS).await` here should be enough to calm down the hot loop

Reply:
I don't think that should work - consider the following code:

```ts
setTimeout(() => { console.log("our code finished!"); }, 5000);
```

then we would never see `our code finished!` because `sleep_for(DEBOUNCE_TIMEOUT_MS)` would be ready prior to our JS code.

Reply:
It wouldn't prevent any execution, just delay it.
Delay can be tweaked, delay_for(0) is good enough to yield 0-10% CPU usage.

Still better to do it explicitly in the hot loop rather than hiding an actual thread sleep somewhere else IMO.

Reply:
I might have misunderstood you. I thought what you suggested was to add `delay_for()` to a branch of `select!` macro, but it seems like my understanding was incorrect.
Where do you propose we should insert `delay_for()`? 

> Also a thread sleep will halt execution, won't it?

I don't think it will. The sleep I inserted to `poll_next` doesn't prevent the tokio's worker thread, and the actual sleep is executed on another thread, which is outside of tokio's thread pool. Once the time has elapsed on the "sleep" thread, it will notify one of the worker threads to poll again.

Reply:
> I thought what you suggested was to add delay_for() to a branch of select!

If none of the futures complete, delay for at-least 0ms.

As is this still runs hot.

![Screenshot 2020-10-26 at 9 12 19 AM](https://user-images.githubusercontent.com/157787/97124517-5d593900-176b-11eb-95a1-8453423fa85a.png)

Reply:
thanks, to be honest I didn't know `tokio::select` has a special branch`else`.
I read `tokio::select` and `futures::select` documents, and it seems like there's a slight difference between tokio::select's `else` branch and futures::select's `default` branch.

[tokio::select - Rust](https://docs.rs/tokio/0.2.22/tokio/macro.select.html)
[futures::select - Rust](https://docs.rs/futures/0.3.7/futures/macro.select.html)

tokio::select's `else` branch is executed only if futures of the other branches have been ready but none of them matches each provided "pattern".
futures::select's `default` branch, on the other hand, is executed if none of the other branches' futures is immediately ready. I think this is what you explained to me.
So, instead of using tokio::select's `else` branch, we should probably add some sleep after the `select!`.